### PR TITLE
Add interface for fantasy observations and efficient cache updates

### PR DESCRIPTION
**Pull request by jacobrgardner**
At long last!

Adds official support for updating an `ExactGP` with fantasy observations:
```python
# Get predictions with original model
test_x = torch.linspace(0, 1, 51)
observed_pred = likelihood(model(test_x))
...
# Get a new model with fantasy observations added to training set, make predictions
fant_model = model.get_fantasy_model(fantasy_x, fantasy_y)
observed_pred = likelihood(fant_model(test_x))
```
The `get_fantasy_model` method returns a copy of `model` with `train_targets`, `train_inputs`, and `prediction_strategy` updated to account for the fantasy data being added to the training set. At the moment, the mean cache is using an efficient bordered linear system solve, rather than computed from scratch. Covariance cache updates are coming soon

## To-do in this PR
- [x] Update covariance cache as well
- [x] Add unit test to test if fantasy model gives same prediction as updating the training data directly with `set_train_data`. (I have this in a notebook, and it passes).
- [x] Add a unit test to ensure we get derivatives w.r.t. fantasy inputs if required, even if `detach_test_caches` is `True`.

## To-do in a later PR
- Add support for adding fantasies to multi-task GPs (really easy, just want to keep the PR small).
- Add support for models that use `InterpolatedLazyTensors`

### CODE REVIEW

**Comment by jacobrgardner** [comment]
The absolute most basic version of this now works as expected and updates both the mean cache and covariance cache (see test.examples.test_simple_gp_regression.test_fantasy_updates) in linear time.

Still a few settings to consider (like multi-task), but let's try to keep this PR minimal and self-contained, and I'll work on those next.

@Balandat @gpleiss

File: gpytorch/models/exact_prediction_strategies.py [hunk 1]
```diff
@@ -73,6 +76,78 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
...
+        full_mean, full_covar = full_output.mean, full_output.lazy_covariance_matrix
+
+        batch_shape = full_inputs[0].shape[:-2]
+
+        full_mean = full_mean.view(*batch_shape, -1)
+        num_train = self.num_train
+
+        # Evaluate fant x train and fant x fant covariance matrices, leave train x train unevaluated.
+        fant_fant_covar = full_covar[..., num_train:, num_train:]
+        mvn = self.likelihood(MultivariateNormal(torch.zeros(1), fant_fant_covar), inputs)
```
Comment (Balandat) [inline gpytorch/models/exact_prediction_strategies.py]: You'll want to use a `.type_as(full_mean)` on the zeros here to make it work on the GPU / with double dtype.  Also, why is this `torch.zeros(1)` and not something like `torch.zeros(*fant_fant_covar.shape[:-2], fant_fant_covar.shape[-1])` ? Or is this all broadcasted appropriately?

File: gpytorch/models/exact_prediction_strategies.py [hunk 2]
```diff
@@ -73,6 +76,78 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
...
+        fant_mean = full_mean[..., num_train:]
+
+        self.fantasy_inputs = inputs
+        self.fantasy_targets = targets
+
+        # Get cached K inverse decomp. (or compute if we somehow don't already have the covariance cache)
+        K_inverse = self.lik_train_train_covar.root_inv_decomposition()
+        fant_solve = K_inverse.matmul(fant_train_covar.transpose(-2, -1))
+
+        schur_complement = fant_fant_covar - fant_train_covar.matmul(fant_solve)
```
Comment (Balandat) [inline gpytorch/models/exact_prediction_strategies.py]: Maybe add a comment or two about what exactly is being computed here since it's not necessarily obvious to everyone. i.e. `K_ff - K_ft K_tt^-1 K_tf`

File: gpytorch/models/exact_prediction_strategies.py [hunk 3]
```diff
@@ -73,6 +76,78 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
...
+        small_system_rhs = targets - fant_mean - fant_train_covar.matmul(self.mean_cache)
+
+        # Schur complement of a spd matrix is guaranteed to be positive definite
+        if small_system_rhs.requires_grad or schur_complement.requires_grad:
+            # TODO: Delete this part of the if statement when PyTorch implements potrs derivative.
```
Comment (Balandat) [inline gpytorch/models/exact_prediction_strategies.py]: Is there an issue open? If so let's add this to the comment.

File: gpytorch/models/exact_prediction_strategies.py [hunk 4]
```diff
@@ -73,6 +76,78 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
...
+        fant_mean_cache = torch.cat((fant_cache_upper, fant_cache_lower), dim=-1)
+
+        # [K U; U S] = [L 0; lower_left schur_root]
+        L_inverse = self.covar_cache
+        L = self.lik_train_train_covar.root_decomposition().root.evaluate()
+        lower_left = fant_train_covar.matmul(L_inverse)
+        schur_root = torch.cholesky(fant_fant_covar - lower_left.matmul(lower_left.transpose(-2, -1)))
+        upper_right = torch.zeros(L.size(-2), schur_root.size(-1))
+
+        # Form pseudo-inverse of [L 0; lower_left schur_root]
```
Comment (Balandat) [inline gpytorch/models/exact_prediction_strategies.py]: instead of concatenating the tensors, how about allocating `new_root` as a zero tensor of appropriate size and then filing the non-zero elements via indexed assignment? Not sure if that makes much of a difference.

File: gpytorch/models/exact_prediction_strategies.py [hunk 5]
```diff
@@ -73,6 +76,78 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
...
+            # TODO: Delete this part of the if statement when PyTorch implements potrs derivative.
+            new_covar_cache = torch.gesv(new_root, cap_mat)[0]
+        else:
+            new_covar_cache = torch.potrs(new_root.transpose(-2, -1), torch.cholesky(cap_mat, upper=True))
+            new_covar_cache = new_covar_cache.transpose(-2, -1)
+
+        # Create new DefaultPredictionStrategy object
+        new_num_train = full_inputs[0].size(len(batch_shape))
+        fant_strat = self.__class__(
+            new_num_train,
```
Comment (Balandat) [inline gpytorch/models/exact_prediction_strategies.py]: mind calling these as kwargs?

File: gpytorch/lazy/lazy_tensor.py [hunk 6]
```diff
@@ -1038,6 +1040,11 @@ def root_inv_decomposition(self, initial_vectors=None, test_vectors=None):
             initial_vectors=initial_vectors,
         )(*self.representation())
 
+        if initial_vectors is not None and initial_vectors.size(-1) > 1:
+            getattr(self, '__cache')["root_decomposition"] = RootLazyTensor(roots[0])
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/lazy/lazy_tensor.py]: Yeah I'm not fully sure why the issue happened but it tried to get the wrong named attribute when calling self.__cache.

File: gpytorch/models/exact_gp.py [hunk 7]
```diff
@@ -61,6 +61,57 @@ def set_train_data(self, inputs=None, targets=None, strict=True):
...
+                               "all test independent caches exist. Call the model on some data first!")
+
+        if self.train_inputs[0].dim() > 2:
+            raise RuntimeError("Adding fantasy observations to a GP that is already batch mode will not be supported "
+                               "until GPyTorch supports multiple batch dimensions.")
+
+        if self.train_targets.dim() > 1:
+            raise RuntimeError("Cannot yet add fantasy observations to multitask GPs, but this is coming soon!")
+
+        if not isinstance(inputs, list):
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/models/exact_gp.py]: No it won't -- this is `inputs = [inputs]` not `inputs = list(inputs)`. This line is here to deal with the case where inputs is a tensor instead of a list.

File: gpytorch/models/exact_gp.py [hunk 8]
```diff
@@ -61,6 +61,57 @@ def set_train_data(self, inputs=None, targets=None, strict=True):
...
+            inputs = [inputs]
+
+        # TODO: Move this code to a new method for re-use with __call__
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/models/exact_gp.py]: Sorry that comment is old, the code in there is no longer really overlapping with call anymore 
```suggestion

```

File: gpytorch/models/exact_prediction_strategies.py (hunk 1 above)
Comment (jacobrgardner) [reply gpytorch/models/exact_prediction_strategies.py]: The mean doesn't much matter but there's  better choice of dummy tensor to use:
```suggestion
        mvn = self.likelihood(MultivariateNormal(fant_mean, fant_fant_covar), inputs)
```

File: gpytorch/models/exact_prediction_strategies.py (hunk 3 above)
Comment (jacobrgardner) [reply gpytorch/models/exact_prediction_strategies.py]: No issue open as far as I know

File: gpytorch/models/exact_gp.py [hunk 9]
```diff
@@ -61,6 +61,57 @@ def set_train_data(self, inputs=None, targets=None, strict=True):
             self.train_targets = targets
         self.prediction_strategy = None
 
+    def get_fantasy_model(self, inputs, targets, **kwargs):
+        if self.prediction_strategy is None:
+            raise RuntimeError("Fantasy observations can only be added after making predictions with a model so that "
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/models/exact_gp.py]: No because creating the fantasy prediction strategy lives inside of prediction strategy, which we don't have or know what it should be without calling the model in eval mode first.

File: test/examples/test_simple_gp_regression.py [hunk 10]
```diff
@@ -159,6 +159,60 @@ def test_posterior_latent_gp_and_likelihood_with_optimization(self, cuda=False):
 
         self.assertLess(mean_abs_error.item(), 0.05)
 
+    def test_fantasy_updates(self, cuda=False):
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply test/examples/test_simple_gp_regression.py]:
```suggestion
    def test_fantasy_updates_cuda(self):
        if torch.cuda.is_available():
            self.test_fantasy_updates(cuda=True)

    def test_fantasy_updates(self, cuda=False):
```

File: test/examples/test_simple_gp_regression.py [hunk 11]
```diff
@@ -159,6 +159,60 @@ def test_posterior_latent_gp_and_likelihood_with_optimization(self, cuda=False):
...
+        # Test the model
+        gp_model.eval()
+        likelihood.eval()
+        test_function_predictions = likelihood(gp_model(test_x))
+
+        # Cut data down, and then add back via the fantasy interface
+        gp_model.set_train_data(train_x[:5], train_y[:5], strict=False)
+        likelihood(gp_model(test_x))
+
+        fantasy_x = torch.nn.Parameter(train_x[5:])
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply test/examples/test_simple_gp_regression.py]:
```suggestion
        fantasy_x = train_x[5:].clone().requires_grad_(True)
```

File: gpytorch/models/exact_gp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Balandat) [reply gpytorch/models/exact_gp.py]: ah right my bad

File: gpytorch/models/exact_prediction_strategies.py (hunk 4 above)
Comment (jacobrgardner) [reply gpytorch/models/exact_prediction_strategies.py]: I think this will mess with autograd as we'll be operating on the tensor in-place. I'll try it though.

File: gpytorch/models/exact_prediction_strategies.py (hunk 2 above)
Comment (jacobrgardner) [reply gpytorch/models/exact_prediction_strategies.py]: Done!

**Comment by jacobrgardner** [comment]
@Balandat Should be all set!

File: gpytorch/functions/_root_decomposition.py [hunk 12]
```diff
@@ -58,11 +60,14 @@ def forward(self, *matrix_args):
             t_mat = t_mat.unsqueeze(0)
         n_probes = t_mat.size(0)
 
-        eigenvalues, eigenvectors = lanczos_tridiag_to_diag(t_mat)
+        mins = NonLazyTensor(t_mat).diag().min(dim=-1, keepdim=True)[0].unsqueeze(-1)
+        jitter_mat = (1e-6 * mins) * torch.eye(t_mat.size(-1)).type_as(t_mat).expand_as(t_mat)
```
Comment (gpleiss) [inline gpytorch/functions/_root_decomposition.py]: Can we parameterize `1e-6`?

File: gpytorch/models/exact_gp.py [hunk 13]
```diff
@@ -61,6 +61,64 @@ def set_train_data(self, inputs=None, targets=None, strict=True):
...
+        train_inputs = [tin.expand(*batch_shape, *tin.shape[-2:]) for tin in list(self.train_inputs)]
+        train_targets = self.train_targets.expand(*batch_shape, *self.train_targets.shape[-2:])
+
+        full_inputs = [torch.cat([train_input, input], dim=-2) for train_input, input in zip(train_inputs, inputs)]
+        full_targets = torch.cat([train_targets, targets], dim=-1)
+
+        full_output = super(ExactGP, self).__call__(*full_inputs, **kwargs)
+
+        # Copy model without copying training data or prediction strategy (since we'll overwrite those)
+        from copy import deepcopy
```
Comment (gpleiss) [inline gpytorch/models/exact_gp.py]: Really picky, but can we move this import up to the top of the file?

File: gpytorch/models/exact_prediction_strategies.py [hunk 14]
```diff
@@ -73,6 +76,116 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
         # where S S^T = (K_XX + sigma^2 I)^-1
         return test_train_covar.matmul(precomputed_cache)
 
+    def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_output):
```
Comment (gpleiss) [inline gpytorch/models/exact_prediction_strategies.py]: Add a docstring here?

File: gpytorch/models/exact_gp.py [hunk 15]
```diff
@@ -61,6 +61,64 @@ def set_train_data(self, inputs=None, targets=None, strict=True):
             self.train_targets = targets
         self.prediction_strategy = None
 
+    def get_fantasy_model(self, inputs, targets, **kwargs):
```
Comment (gpleiss) [inline gpytorch/models/exact_gp.py]: Add a docstring here?

**Comment by jacobrgardner** [comment]
@gpleiss Done here, too!
